src/main/java/com/example/hop_oasis/service/data/RecommendationsServiceImpl.java

```diff
@@ -0,0 +1,185 @@
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+
+@Service
+public class RecommendationsServiceImpl implements RecommendationsService {
+    private final BeerRepository beerRepository;
+    private final CiderRepository ciderRepository;
+    private final SnackRepository snackRepository;
+    private final ProductBundleRepository productBundleRepository;
+    private final CartRepository cartRepository;
+
+    public RecommendationsServiceImpl(BeerRepository beerRepository,
+                                      CiderRepository ciderRepository,
+                                      SnackRepository snackRepository,
+                                      ProductBundleRepository productBundleRepository,
+                                      CartRepository cartRepository
+    ) {
+        this.beerRepository = beerRepository;
+        this.ciderRepository = ciderRepository;
+        this.snackRepository = snackRepository;
+        this.productBundleRepository = productBundleRepository;
+        this.cartRepository = cartRepository;
+    }
+
+    @Override
+    public Recommendations getForCart(Long cartId) {
+        final var cart = cartRepository.findById(cartId)
+                .orElseThrow(() -> new ResourceNotFoundException("Cart not found, id: %s", cartId));
+
+        final var beerSet = new HashSet<Beer>();
+        final var ciderSet = new HashSet<Cider>();
+        final var snacksSet = new HashSet<Snack>();
+        final var bundleSet = new HashSet<ProductBundle>();
+
+        for (CartItem cartItem : cart.getCartItems()) {
+            final var recommendations = getForProduct(cartItem.getItemId(), cartItem.getItemType().name());
+            beerSet.addAll(recommendations.getBeers());
+            ciderSet.addAll(recommendations.getCiders());
+            snacksSet.addAll(recommendations.getSnacks());
+            bundleSet.addAll(recommendations.getBundles());
+        }
+
+        // remove what already in cart
+        for (CartItem cartItem : cart.getCartItems()) {
+            switch (cartItem.getItemType()) {
+                case BEER -> beerSet.removeIf(beer -> beer.getId().equals(cartItem.getItemId()));
+                case CIDER -> ciderSet.removeIf(cider -> cider.getId().equals(cartItem.getItemId()));
+                case SNACK -> snacksSet.removeIf(snack -> snack.getId().equals(cartItem.getItemId()));
+                case PRODUCT_BUNDLE -> bundleSet.removeIf(pb -> pb.getId().equals(cartItem.getItemId()));
+                case null -> throw new IllegalStateException("Item type is null");
+            }
+        }
+
+        return new Recommendations(
+                getFirstN(beerSet.stream().toList(), 5),
+                getFirstN(ciderSet.stream().toList(), 5),
+                getFirstN(snacksSet.stream().toList(), 5),
+                getFirstN(bundleSet.stream().toList(), 5));
```

Comment:
You should limit number of records retrieved from the DB?
You can use PageRequest.of(...)